### PR TITLE
Bridge command generation to powershell when on Windows

### PR DESCRIPTION
**Pull request by eddy-win**
## What? Why? How?
- When running on Windows, codex often tries to invoke bash commands, which commonly fail (unless WSL is installed)
- Fix: Detect if powershell is available and, if so, route commands to it
  - Also add a shell_name property to environmental context for codex to default to powershell commands when running in that environment

## Testing
- Tested within WSL and powershell (e.g. get top 5 largest files within a folder and validated that commands generated were powershell commands)
- Tested within Zsh
- Updated unit tests

### CODE REVIEW

**Comment by eddy-win** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by eddy-win** [comment]
recheck

File: codex-rs/core/src/environment_context.rs [hunk 1]
```diff
@@ -28,13 +28,15 @@ pub(crate) struct EnvironmentContext {
...
     pub network_access: NetworkAccess,
+    pub shell_name: Option<String>,
 }
 
 impl EnvironmentContext {
     pub fn new(
         cwd: PathBuf,
         approval_policy: AskForApproval,
         sandbox_policy: SandboxPolicy,
+        shell_name: Option<String>,
```
Comment (bolinfest) [inline codex-rs/core/src/environment_context.rs]: If convenient, can we pass `Shell` instead of `String`?

File: codex-rs/core/src/shell.rs [hunk 2]
```diff
@@ -233,3 +291,34 @@ mod tests {
...
+
+        let out = shell.format_default_shell_invocation(input);
+
+        assert_eq!(
+            out,
+            Some(vec![
+                "pwsh.exe".to_string(),
+                "-NoProfile".to_string(),
+                "-Command".to_string(),
+                "echo hello".to_string()
```
Comment (bolinfest) [inline codex-rs/core/src/shell.rs]: Arbitrary bash commands won't work though, will they?

Comment (bolinfest) [reply codex-rs/core/src/shell.rs]: Should we be checking for `git-bash` instead in this case?

File: codex-rs/core/src/codex.rs [hunk 3]
```diff
@@ -1795,11 +1796,14 @@ pub struct ExecInvokeArgs<'a> {
 }
 
 fn maybe_run_with_user_profile(params: ExecParams, sess: &Session) -> ExecParams {
-    if sess.shell_environment_policy.use_profile {
-        let command = sess
+    let should_translate = matches!(sess.user_shell, crate::shell::Shell::PowerShell(_))
```
Comment (bolinfest) [inline codex-rs/core/src/codex.rs:2061]: I think we need to rename the function if we stick with this...

File: codex-rs/core/src/shell.rs (hunk 2 above)
Comment (eddy-win) [reply codex-rs/core/src/shell.rs]: you're right that general commands don't work. things like "ls" and "rm" DO work. Once nice thing of forcing powershell was that, if the model hadn't realized it should use powershell, yet, the error output was a further clue it should switch to powershell. I'll add detection for git-bash and even bash. Regular bash itself CAN be present within windows host if WSL has been enabled.

**Comment by eddy-win** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by eddy-win** [comment]
recheck

**Comment by eddy-oai** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by eddy-oai** [comment]
recheck

File: codex-rs/core/src/shell.rs (hunk 2 above)
Comment (eddy-oai) [reply codex-rs/core/src/shell.rs]: added a fallback to bash, if we can find a bash executable

File: codex-rs/core/src/shell.rs [hunk 4]
```diff
@@ -1,14 +1,23 @@
...
 pub struct ZshShell {
     shell_path: String,
     zshrc_path: String,
 }
 
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct PowerShellShell {
+    exe: String, // Executable name or path, e.g. "pwsh" or "powershell.exe".
+    bash_exe_fallback: Option<String>, // In case the model generates a bash command.
```
Comment (bolinfest) [inline codex-rs/core/src/shell.rs]: Should this be `Option<PathBuf>`?

File: codex-rs/core/src/shell.rs [hunk 5]
```diff
@@ -1,14 +1,23 @@
...
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
 pub struct ZshShell {
     shell_path: String,
     zshrc_path: String,
 }
 
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct PowerShellShell {
```
Comment (bolinfest) [inline codex-rs/core/src/shell.rs]: Maybe `PowerShellConfig` would be less confusing?

File: codex-rs/core/src/shell.rs [hunk 6]
```diff
@@ -33,6 +42,58 @@ impl Shell {
...
+
+                    // No bash fallback → run the script under PowerShell.
+                    // It will likely fail (except for some simple commands), but the error
+                    // should give a clue to the model to fix upon retry that it's running under PowerShell.
+                    return Some(vec![
+                        ps.exe.clone(),
+                        "-NoProfile".to_string(),
+                        "-Command".to_string(),
+                        script,
+                    ]);
```
Comment (bolinfest) [inline codex-rs/core/src/shell.rs]: The more canonical way to write this in Rust would be:

```suggestion
                    return match &ps.bash_exe_fallback {
                        Some(bash) => Some(vec![bash.clone(), "-lc".to_string(), script]),
                        None => {

                    // No bash fallback → run the script under PowerShell.
                    // It will likely fail (except for some simple commands), but the error
                    // should give a clue to the model to fix upon retry that it's running under PowerShell.
                    Some(vec![
                        ps.exe.clone(),
                        "-NoProfile".to_string(),
                        "-Command".to_string(),
                        script,
                    ])
                }
            }
```

File: codex-rs/core/src/shell.rs [hunk 7]
```diff
@@ -33,6 +42,58 @@ impl Shell {
...
+                }
+
+                // Not a bash command. If model did not generate a PowerShell command,
+                // turn it into a PowerShell command.
+                let first = command.first().map(String::as_str);
+                if first != Some(ps.exe.as_str()) {
+                    // TODO (CODEX_2900): Handle escaping newlines.
+                    if command.iter().any(|a| a.contains('\n') || a.contains('\r')) {
```
Comment (bolinfest) [inline codex-rs/core/src/shell.rs:73]: This seems a bit dicey?

File: codex-rs/core/src/shell.rs [hunk 8]
```diff
@@ -33,6 +42,58 @@ impl Shell {
...
+                        return Some(command);
+                    }
+
+                    let joined = shlex::try_join(command.iter().map(|s| s.as_str())).ok();
```
Comment (bolinfest) [inline codex-rs/core/src/shell.rs:77]: Is shlex's quoting rules appropriate for Windows?

File: codex-rs/core/src/shell.rs [hunk 9]
```diff
@@ -33,6 +42,58 @@ impl Shell {
...
+                    if let Some(joined) = joined {
+                        return Some(vec![
+                            ps.exe.clone(),
+                            "-NoProfile".to_string(),
+                            "-Command".to_string(),
+                            joined,
+                        ]);
+                    }
+                    return None;
```
Comment (bolinfest) [inline codex-rs/core/src/shell.rs]: Can probably do something like:

```suggestion
                    return joined.map(|arg| {
                        vec![
                            ps.exe.clone(),
                            "-NoProfile".to_string(),
                            "-Command".to_string(),
                            arg,
                        ]
                    };
```

File: codex-rs/core/src/shell.rs [hunk 10]
```diff
@@ -86,11 +147,51 @@ pub async fn default_user_shell() -> Shell {
...
 pub async fn default_user_shell() -> Shell {
     Shell::Unknown
 }
 
+#[cfg(target_os = "windows")]
+pub async fn default_user_shell() -> Shell {
+    use tokio::process::Command;
+
+    // Prefer PowerShell 7+ (`pwsh`) if available, otherwise fall back to Windows PowerShell.
+    let has_pwsh = Command::new("pwsh")
```
Comment (bolinfest) [inline codex-rs/core/src/shell.rs:164]: Consider `tokio::join!` to run these in parallel since this is on the startup path via `Session::new()`?

**Review by bolinfest — APPROVED** [review]
Thanks, overall this seems great! I do worry a bit about the escaping in the one place I called out. Could we at least get more unit tests to try to flag edge cases like newlines and other types of whitespace that might make parsing funny?

File: codex-rs/core/src/shell.rs (hunk 7 above)
Comment (eddy-oai) [reply codex-rs/core/src/shell.rs:73]: Yes, flagged CODEX-2900 for follow-up

File: codex-rs/core/src/shell.rs (hunk 8 above)
Comment (eddy-oai) [reply codex-rs/core/src/shell.rs:77]: It is not the best, flagged in CODEX-2900 for follow-up
